### ISSUE TITLE
reMarkable 2 support

### ISSUE THREAD

**Issue by fedecupe** [matteodelabre/vnsee#15]
Hi there,
Any hope for remarkable 2 support?

Thanks for your work.

**Comment by matteodelabre** [comment]
Interacting with the screen of the reMarkable 2 turns out to be considerably harder than with the reMarkable 1, since it is not exposed as a Linux framebuffer device. There is an ongoing reverse-engineering effort to achieve that in the <https://github.com/ddvk/remarkable2-framebuffer> repository. Once it stabilizes and we have enough people able to test it on their rM2 tablets (which I don’t currently own), I’ll be happy to work on integrating it into VNSee!

**Comment by fedecupe** [comment]
Thanks, if you need some debug I have the rm2 and have a Linux Box.

**Comment by ennioVisco** [comment]
I'll be glad to help testing it! Unfortunately, I'm on Remarkable 2, and according to #13 there is still no stable solution for printing on Remarkable 2. I'll be waiting for that and than come back 😊

**Comment by LeifAndersen** [comment]
For what its worth I've actually gotten this to work on the remarkable2, using the remarkable2-framebuffer repo.

https://www.youtube.com/watch?v=X0_6_c_1kBE&feature=youtu.be

**Comment by LeifAndersen** [comment]
Obviously the refresh rate is to slow to watch a video....but for reading websites it works pretty great.

**Comment by ennioVisco** [comment]
> For what its worth I've actually gotten this to work on the remarkable2, using the remarkable2-framebuffer repo.
> 
> https://www.youtube.com/watch?v=X0_6_c_1kBE&feature=youtu.be

Any changes, or it works just out of the box?

**Comment by LeifAndersen** [comment]
I did not need to modify vnsee or remarkable2-framebuffer. But I did compile both of them myself using the rm2 sdk.

Basically, I opened up two shells over ssh. First I shut down xochitl:
```
systemctl shutdown xochitl
```

Then, in the first shell I ran:
```
LD_PRELOAD="./dist/librm2fb_server.so.1.0.0" /usr/bin/remarkable-shutdown
```

Then, I started a vnc server on my desktop:
```
x11vnc -repeat -forever -nocursor -allow 10.11.99.1 -nopw -geometry 1872x1404 -rotate -90
```

Finally, in the second shell I ran:
```
LD_PRELOAD="./dist/librm2fb_client.so.1.0.0" ./vnsee
```

**Comment by Equator11111** [comment]
This worked for me as well. However I was unable to get input working (both from pen and touch). Is this something that will still have to be reworked for the RM2 or is there something that I am missing?

**Comment by LeifAndersen** [comment]
Very strange. I'm also noticing the same behaviour, which is odd since the rm2 seems to use similar input to the rm1. Maybe its possible you hard-coded some device ids that changed?

I'll have to look into this more this weekend.

**Comment by NickHu** [comment]
It's probably this https://github.com/Eeems/oxide/issues/48#issuecomment-690830572

**Comment by o-** [comment]
> It's probably this Eeems/oxide#48 (comment)

I tried to just change these inputs and compile. It's going in the right direction, though there is more to it. It seems the resulting cursor location is quite off. And I can scroll down, but not up :)

**Comment by matteodelabre** [comment]
Thanks all for your input! I’ve recently bought a rM2 and I’m currently working on adding rM2 support.

**Comment by matteodelabre** [comment]
Hello everyone, I’m glad to announce that rM2 support has reached a beta state. Here’s a development build for those of you who would like to test it: [vnsee-rm2-beta-6088f4.zip](https://github.com/matteodelabre/vnsee/files/5958781/vnsee-rm2-beta-6088f4.zip). (You can also build it from the [`rm2-support` branch](https://github.com/matteodelabre/vnsee/tree/rm2-support) if you’d like to inspect the source by yourself.)

Some notes:

* You’ll need a working install of the [remarkable2-framebuffer](https://github.com/ddvk/remarkable2-framebuffer) project for VNSee to work. You can install it easily from [Toltec](https://toltec-dev.org). If you see the error message below, you have launched VNSee without a running rm2fb server:

```
Error: (rmioc::screen_rm2fb) Open shared memory framebuffer: No such file or directory
```

* You do not need to launch `vnsee` with `LD_PRELOAD="…/librm2fb_client.so.1.0.0" vnsee` or `rm2fb-client vnsee`, just type `./vnsee` directly.
* The VNC server that you connect to must have a resolution of 1404x3840, instead of 1408x3840 (this is due to differences in the underlying screen buffer). This difference will be fixed in #11.
* The full refresh option, triggered by pressing the home key on the reMarkable 1, is not available since there is no such key on the reMarkable 2. This will probably be implemented later on through a touch gesture.

If you test this beta build, please feel free to share any remarks or bugs in this thread and I’ll make sure to address them.

**Comment by kunalb** [comment]
I was trying to test it out today, but can't unzip the development build:
```
   unzip vnsee-rm2-beta-018b42.zip
Archive:  vnsee-rm2-beta-018b42.zip
  End-of-central-directory signature not found.  Either this file is not
  a zipfile, or it constitutes one disk of a multi-part archive.  In the
  latter case the central directory and zipfile comment will be found on
  the last disk(s) of this archive.
unzip:  cannot find zipfile directory in one of vnsee-rm2-beta-018b42.zip or
        vnsee-rm2-beta-018b42.zip.zip, and cannot find vnsee-rm2-beta-018b42.zip.ZIP, period.
```

And fetching install-toolchain is instead redirecting me to the remarkable website, so I haven't been able to manually build it yet :(. Could you upload a new copy of the build?

**Comment by matteodelabre** [comment]
> I was trying to test it out today, but can't unzip the development build

Sorry about that! It actually was a tar archive instead of a zip. Here’s the same archive in the zip format: [vnsee-rm2-beta-018b42.zip](https://github.com/matteodelabre/vnsee/files/5961668/vnsee-rm2-beta-018b42.zip).

> And fetching install-toolchain is instead redirecting me to the remarkable website, so I haven't been able to manually build it yet :(. Could you upload a new copy of the build?

Unfortunately the company removed the toolchain from its website a few weeks ago for reasons that are still unknown. It’s still available through the Internet Archive though: https://web.archive.org/web/20201129102245/https://remarkable.engineering/oecore-x86_64-cortexa9hf-neon-toolchain-zero-gravitas-1.8-23.9.2019.sh

**Comment by kunalb** [comment]
Thanks for both! Could successfully tar -xf the original build.

**Comment by kunalb** [comment]
Just to confirm, did you mean that the resolution should be 1404x1872? 

At the moment I get an error of failing to initialize the VNC connection after it tries for several seconds.

**Comment by matteodelabre** [comment]
> Just to confirm, did you mean that the resolution should be 1404x1872?

Yup, thanks! I fixed it.

> At the moment I get an error of failing to initialize the VNC connection after it tries for several seconds.

Is your VNC server running on port 5900?

**Comment by kunalb** [comment]
Yup -- the x11vnc output explicitly prints out PORT=5900.

**Comment by matteodelabre** [comment]
Is there any output from `x11vnc` when you start `vnsee`?

**Comment by kunalb** [comment]
No, nothing from x11vnc. I'm not quite sure how to debug further.

**Comment by matteodelabre** [comment]
If there’s no output from *x11vnc* then it looks like a mismatch between the VNC server’s address and what address VNSee is trying to reach. When you start VNSee, you should see a first message saying `Connecting to 10.11.99.x:5900…`. Can you check that your computer’s IP on the USB network (which can be seen in the output of `ip addr`) matches with what VNSee shows?

**Comment by fedecupe** [comment]
Hi, I get this error when trying to launch vnsee on remarkable2

```
reMarkable: ~/ ./vnsee
Connecting to 10.11.99.4:5900...
terminate called after throwing an instance of 'std::runtime_error'
  what():  Server uses an unsupported resolution (1408x1872). This client can only cope with a screen width of exactly 260 pixels and a screen height of 23936 pixels
Aborted (core dumped)
```

**Comment by matteodelabre** [comment]
Hi! Thanks for the report. Could you please check whether `cat /sys/devices/soc0/machine` reports `reMarkable 2.0`? Also, just to be sure, could you double-check that you’re running the dev build and not an old one? (`sha256sum vnsee` should print the following hash: `35138600893ef76f5775ec263704f811341e2e4c31b915da9602e323392d98fe`)

**Comment by kunalb** [comment]
Hmm, ip addr gives me the right addresses and things *look* to be sane on both sides. (10.11.99.2)
But I can't actually hit any port from the remarkable to the laptop: I tried to telnet into the vnc server, just to see if it worked and remove a variable and that didn't work. I also tried to hit an http server on my laptop from remarkable and that also failed.

ping works.

**Comment by matteodelabre** [comment]
Can you check whether you have an iptables rule filtering inbound traffic on the USB network interface? Is there any output from `dmesg -w` on your computer when you attempt to telnet from the rM?

**Comment by kunalb** [comment]
Looks like it's iptables rule filtering set up for my office laptop; looking into that now! Thank you for pointing me to that!
dmesg -w didn't have any output.

**Comment by kunalb** [comment]
All right, forcefully ACCEPTING packets from USB let me start connecting! With that in place, I now get an error for
```
reMarkable: ~/ ./vnsee
Connecting to 10.11.99.2:5900...
terminate called after throwing an instance of 'std::runtime_error'
  what():  Server uses an unsupported resolution (1408x1872). This client can only cope with a screen width of exactly 1404 pixels and a screen height of 1872 pixels
Aborted (core dumped)
```

maybe because of
```
gtf 1404 1872 60

  # 1408x1872 @ 60.00 Hz (GTF) hsync: 116.22 kHz; pclk: 225.00 MHz
  Modeline "1408x1872_60.00"  225.00  1408 1520 1672 1936  1872 1873 1876 1937  -HSync +Vsync
```

I haven't dug into this before; should I manually try replacing 1408 by 1404 while creating the new mode?

**Comment by matteodelabre** [comment]
> All right, forcefully ACCEPTING packets from USB let me start connecting!

Great!

> I haven't dug into this before; should I manually try replacing 1408 by 1404 while creating the new mode?

`gtf` automatically rounds arguments to be a multiple of 8 it seems. Actually we don’t really need to bother with the hsync/vsync arguments of the modeline since it’s a virtual output, so you could simply use:

```
xrandr --newmode 1404x1872 0 1404 1404 1404 1404 1872 1872 1872 1872
```

I’ll update the docs accordingly.

**Comment by kunalb** [comment]
Keeping the 1408 and only changing the clip on VNC let me start things working! I'll try with the simpler mode definition as well, thank you so much! I'm tying this out on the remarkable screen now.

**Comment by Baccanno** [comment]
So used the https://github.com/matteodelabre/vnsee/files/5961668/vnsee-rm2-beta-018b42.zip after having toltec installed rm2fb and I manage to reach this step
```
./vnsee
Connecting to 10.11.99.2:5900...
Error: Failed to initialize VNC connection
```

I believe I time out (after 1min or so) because nothing connects to it. I'm on windows and I'll have to look into available VNC servers.

**Comment by matteodelabre** [comment]
Thanks for testing @Baccanno! I will be publishing a release including rM2 support soon (probably in the coming week).

**Comment by lapwat** [comment]
# Workaround instructions

## Mono display

@LeifAndersen [instructions](https://github.com/matteodelabre/vnsee/issues/15#issuecomment-751760571) work great

## Multiple displays

```sh
# host shell
wget https://github.com/ddvk/remarkable2-framebuffer/releases/download/v0.0.3/librm2fb_client.so.1.0.0
scp librm2fb_client.so.1.0.0 root@$REMARKABLE_IP:dist # dist directory should exist

xrandr --newmode 1872x1404_landscape 0 1872 1872 1872 1872 1404 1404 1404 1404
xrandr --addmode VIRTUAL1 1872x1404
xrandr --output VIRTUAL1 --mode 1872x1404 --right-of eDP1
x11vnc -repeat -forever -nocursor -allow $REMARKABLE_IP -nopw -clip 1872x1404+1920+0% -rotate -90

# remarkable shell
systemctl stop xochitl
LD_PRELOAD="./dist/librm2fb_server.so.1.0.0" /usr/bin/remarkable-shutdown
LD_PRELOAD="./dist/librm2fb_client.so.1.0.0" ./vnsee
```

**Comment by matteodelabre** [comment]
Just released [version 0.4.0](https://github.com/matteodelabre/vnsee/releases/tag/v0.4.0) with support for reMarkable 2. Thanks to everyone who provided feedback on the development build!
